File: src/scrapers/bunert.py
"""Scraper for Bunert.de Blackroll product(s)."""

from __future__ import annotations

from datetime import datetime, UTC
from urllib.parse import urljoin

from playwright.async_api import Page

from ..models import Product, ScrapeResult
from .base import BaseScraper
from .browser_pool import get_browser_context


_DESKTOP_UA = (
    "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 "
    "(KHTML, like Gecko) Chrome/125.0.0.0 Safari/537.36"
)

_HIDE_WEBDRIVER = """
Object.defineProperty(navigator, 'webdriver', { get: () => undefined });
"""


class BunertScraper(BaseScraper):
    name = "bunert"
    display_name = "Bunert"
    url = "https://www.bunert.de/blackroll-standard-faszienrolle"

    async def scrape(self) -> ScrapeResult:
        async with get_browser_context(
            user_agent=_DESKTOP_UA,
            locale="de-DE",
            viewport={"width": 1400, "height": 900},
            init_scripts=[_HIDE_WEBDRIVER],
        ) as context:
            page = await context.new_page()
            print(f"[{self.name}] Loading {self.url}...")
            await page.goto(self.url, wait_until="domcontentloaded", timeout=90000)
            await page.wait_for_timeout(400)

            products = await self.extract_products(page)
            print(f"[{self.name}] Extracted {len(products)} products")

        return ScrapeResult(
            source=self.name,
            source_url=self.url,
            scraped_at=datetime.now(UTC),
            products=products,
        )

    @staticmethod
    async def _data_layer_products(page: Page) -> tuple[list[dict], str | None]:
        data_layer = await page.evaluate("() => window.dataLayer || []")
        if not isinstance(data_layer, list):
            return [], None

        for entry in data_layer:
            if not isinstance(entry, dict):
                continue
            ecommerce = entry.get("ecommerce")
            if not isinstance(ecommerce, dict):
                continue
            detail = ecommerce.get("detail")
            if not isinstance(detail, dict):
                continue
            products = detail.get("products")
            if isinstance(products, list) and products:
                currency = ecommerce.get("currencyCode")
                return [p for p in products if isinstance(p, dict)], currency if isinstance(currency, str) else None
        return [], None

    @staticmethod
    async def _fetch_image(page: Page, image_url: str) -> tuple[bytes | None, str | None]:
        headers = {
            "Accept": "image/avif,image/webp,image/*,*/*;q=0.8",
            "Referer": page.url,
            "Sec-Fetch-Dest": "image",
            "Sec-Fetch-Mode": "no-cors",
        }
        resp = await page.context.request.get(image_url, timeout=45000, headers=headers)
        if not resp.ok:
            return None, None
        body = await resp.body()
        if len(body) < 800:
            return None, None
        mime = resp.headers.get("content-type")
        if mime and ";" in mime:
            mime = mime.split(";", 1)[0].strip()
        return body, mime

    async def extract_products(self, page: Page) -> list[Product]:
        dl_products, currency = await self._data_layer_products(page)

        name = None
        item_id = None
        price = None

        if dl_products:
            p0 = dl_products[0]
            if isinstance(p0.get("name"), str):
                name = p0["name"].strip() or None
            if isinstance(p0.get("id"), str):
                item_id = p0["id"].strip() or None
            raw_price = p0.get("price")
            try:
                if raw_price is not None:
                    price = float(str(raw_price).replace(",", "."))
            except Exception:
                price = None

        if not currency:
            currency = "EUR"

        if not name:
            try:
                title = await page.title()
                name = (title or "").split("|")[0].strip() or None
            except Exception:
                name = None

        image_url = await page.get_attribute('meta[property="og:image"]', "content")
        if image_url:
            image_url = urljoin(page.url, image_url)

        image_bytes = None
        image_mime = None
        if image_url:
            try:
                image_bytes, image_mime = await self._fetch_image(page, image_url)
            except Exception as e:
                print(f"[{self.name}] Failed to fetch image: {e}")

        if not name:
            return []

        return [
            Product(
                name=name,
                price=price,
                currency=currency,
                url=page.url,
                item_id=item_id,
                image=image_bytes,
                image_mime=image_mime,
            )
        ]

File: new/packages/scrapers/src/bunert.ts
import type { DiscoveredProduct } from "@corastuff/shared";
import type { Page } from "playwright";
import { withPlaywrightContext, type PlaywrightContextProfile, type PlaywrightRunArtifactsOptions } from "./playwrightContext.js";

function asNonEmptyString(value: unknown): string | undefined {
  return typeof value === "string" && value.trim() ? value.trim() : undefined;
}

function normalizeUrl(baseUrl: string, value: string | undefined): string | undefined {
  const raw = asNonEmptyString(value);
  if (!raw) return undefined;
  if (raw.startsWith("//")) return `https:${raw}`;
  if (raw.startsWith("/")) return new URL(raw, baseUrl).toString();
  try {
    return new URL(raw, baseUrl).toString();
  } catch {
    return raw;
  }
}

function coercePrice(value: unknown): number | undefined {
  if (typeof value === "number" && Number.isFinite(value) && value > 0) return value;
  if (typeof value === "string") {
    const raw = value.trim();
    if (!raw) return undefined;
    const cleaned = raw.replace(/[^\d,.\-]/g, " ").replace(/\s+/g, " ").trim();
    const match = cleaned.match(/-?\d{1,10}(?:[.,]\d{1,2})?/);
    if (!match) return undefined;
    const token = match[0]!;
    const hasComma = token.includes(",");
    const hasDot = token.includes(".");
    const normalized =
      hasComma && hasDot ? token.replace(/\./g, "").replace(",", ".") : hasComma ? token.replace(",", ".") : token;
    const num = Number.parseFloat(normalized);
    return Number.isFinite(num) && num > 0 ? num : undefined;
  }
  return undefined;
}

const desktopUserAgent =
  "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/131.0.0.0 Safari/537.36";

type DataLayerProduct = { name?: unknown; id?: unknown; price?: unknown };

async function extractFromDataLayer(page: Page): Promise<{
  name?: string;
  itemId?: string;
  price?: number;
  currency?: string;
}> {
  const result: unknown = await page.evaluate(() => {
    const dataLayer = (globalThis as any)?.window?.dataLayer ?? (globalThis as any)?.dataLayer ?? [];
    if (!Array.isArray(dataLayer)) return null;

    for (const entry of dataLayer) {
      const ecommerce = entry?.ecommerce;
      const detail = ecommerce?.detail;
      const products = detail?.products;
      if (!Array.isArray(products) || products.length === 0) continue;
      const first = products.find((p) => typeof p === "object" && p !== null && !Array.isArray(p)) ?? products[0];
      return {
        product: first,
        currencyCode: typeof ecommerce?.currencyCode === "string" ? ecommerce.currencyCode : null
      };
    }

    return null;
  });

  if (typeof result !== "object" || result === null) return {};
  const obj = result as { product?: unknown; currencyCode?: unknown };

  const productRaw = obj.product;
  const product =
    typeof productRaw === "object" && productRaw !== null && !Array.isArray(productRaw)
      ? (productRaw as DataLayerProduct)
      : undefined;

  const name = product ? asNonEmptyString(product.name) : undefined;
  const itemId = product ? asNonEmptyString(product.id) : undefined;
  const price = product ? coercePrice(product.price) : undefined;
  const currency = asNonEmptyString(obj.currencyCode)?.toUpperCase();

  return {
    ...(name ? { name } : {}),
    ...(itemId ? { itemId } : {}),
    ...(price !== undefined ? { price } : {}),
    ...(currency ? { currency } : {})
  };
}

async function safeGetMeta(page: Page, selector: string): Promise<string | undefined> {
  try {
    return asNonEmptyString(await page.locator(selector).first().getAttribute("content"));
  } catch {
    return undefined;
  }
}

export async function scrapeBunertProductPage(options: {
  sourceSlug: string;
  productUrl: string;
  currency?: string;
  browser?: PlaywrightContextProfile;
  artifacts?: PlaywrightRunArtifactsOptions;
  userAgent?: string;
  locale?: string;
  log?: (message: string) => void;
}): Promise<{ products: DiscoveredProduct[] }> {
  const userAgent = options.browser?.userAgent ?? options.userAgent ?? desktopUserAgent;
  const locale = options.browser?.locale ?? options.locale ?? "de-DE";
  const browser: PlaywrightContextProfile = {
    ...options.browser,
    userAgent,
    locale,
    viewport: options.browser?.viewport ?? { width: 1400, height: 900 },
    stealth: options.browser?.stealth ?? true
  };

  return await withPlaywrightContext(
    browser,
    async (context) => {
    const page = await context.newPage();
    try {
      options.log?.(`Loading ${options.productUrl}`);
      await page.goto(options.productUrl, { waitUntil: "domcontentloaded", timeout: 90_000 });
      await page.waitForTimeout(400);

      const baseUrl = (() => {
        try {
          return new URL(options.productUrl).origin;
        } catch {
          return "https://www.bunert.de/";
        }
      })();

      const fromDl = await extractFromDataLayer(page);
      const titleFallback = (() => {
        try {
          return page?.title?.();
        } catch {
          return null;
        }
      })();

      const rawTitle = titleFallback ? await titleFallback : undefined;
      const titleName = asNonEmptyString(rawTitle)?.split("|")[0]?.trim();
      const name = fromDl.name ?? titleName;
      if (!name) return { products: [] };

      const url = page.url();
      const imageUrl = normalizeUrl(baseUrl, await safeGetMeta(page, 'meta[property="og:image"]'));

      const currency =
        asNonEmptyString(fromDl.currency) ??
        asNonEmptyString(options.currency)?.toUpperCase() ??
        (fromDl.price !== undefined ? "EUR" : undefined);

      return {
        products: [
          {
            sourceSlug: options.sourceSlug,
            itemId: fromDl.itemId,
            name,
            url,
            price: fromDl.price,
            currency: fromDl.price !== undefined ? currency : undefined,
            imageUrl
          }
        ]
      };
    } finally {
      await page.close().catch(() => undefined);
    }
    },
    { artifacts: options.artifacts }
  );
}
